### DISCUSSION TITLE
커뮤니티에 어떤 기능을 추가하면 좋을까요?

### DISCUSSION THREAD

**Discussion by hgo641** [woowacourse-teams/2023-hang-log#539] [토론]
팀 회의 결과, 행록에 커뮤니티를 추가하기로 했습니다. 🎉
커뮤니티의 컨셉과 상세 기능에 대한 토론이 필요할 것 같아 디스커션을 열게 되었습니다.
두 개의 안건을 준비했습니다! 
**의견을 작성할 때, 왜 그렇게 생각하는지 이유도 함께 남겨주세요!**

### 안건 1 : 게시물의 공유 범위를 어떻게 잡을 것인가?
1. 모든 사용자에게 게시물을 공유한다.
2. 링크를 직접 공유한 사용자에게만 게시물을 공유한다.
3. 1번과 2번의 기능을 전부 제공한다.
4. 그 외의 방식

### 안건 2 : 커뮤니티에 어떤 기능을 추가하면 좋을까?
> 예시
> * 게시물 좋아요 기능
> * 게시물 조회수 표시 기능
> * 게시물 댓글 기능

**Comment by jjongwa** [comment]
### 안건 1
전 `3. 1번과 2번의 기능을 전부 제공한다.` 입장인데요.
전체에게 공개가 아닌 특정 인물에게 여행을 공유하고 싶어할 수도 있기 때문에
`공개 - 비공개` `공유-비공유` 버튼을 따로 두어 관리하는게 좋다고 생각헀습니다.

### 안건 2
제가 생각한 부가 기능들은 다음과 같습니다.
- 게시물 좋아요 기능
- 게시물 조회수 표시 기능 - 기본(클릭 당 조회수가 올라가는 방식)
- 게시물 조회수 표시 기능 - 심화(하루에 한 유저가 여러번 조회해도 count는 +1)
- 게시물 댓글 기능(대댓글)
- 게시물 댓글에 대한 알림 기능
- 게시물 검색 기능(제목별, 도시별)
- 게시물 추천 기능(가장 인기있는, 최근에 올라온 등등)
하지만 이러한 기능들 중에 현실적으로 가능한 정도는
`게시물 좋아요 기능`과 `게시물 조회수 표시 기능 - 기본` 이 두가지라고 생각하고, 
해당 기능을 구현 완료한 뒤에는 `게시물 검색 & 추천` 기능 을 도전해 볼 수 있을 것 같습니다.
검색과 추천 모두 백엔드에서 로직으로만 구현하고 프론트 단에서는 원래 보여주던 게시물들의 순서를 바꾸어 보여주는 것이기 때문에 저희 쪽만 시간이 된다면..? 해 볼만한 기능이 아닐까 싶네요

**Comment by hgo641** [comment]
### 안건 1 : 게시물의 공유 범위를 어떻게 잡을 것인가?
`3. 1번과 2번의 기능을 전부 제공한다.` 가 좋은 것 같아요!
이유는 디노랑 같습니다.ㅎㅎ 
특정 사용자에게 공유하는 기능은 현재 구현되어 있는 공유 링크 기능으로 유지하고,
모든 사용자가 볼 수 있는 전체 커뮤니티를 추가하면 좋을 것 같아요!

### 안건 2 : 커뮤니티에 어떤 기능을 추가하면 좋을까?
* 게시물 조회수 표시 기능(게시물 클릭 당 조회수 상승)
* 게시물 좋아요 기능
* 게시물 좋아요 알림 기능
* 게시물 검색 기능(제목, 도시, 사용자)
    * 제목란에 문자열을 입력하면 정확도순으로 조회
    * 도시란에서 특정 도시들을 선택하면 필터링 가능
    * 사용자란에서 문자열을 입력하면 정확도순으로 조회
* 게시물 신고 기능 

조회수랑 좋아요 볼 수 있으면 저처럼 관심을 좋아하는 유저에게 너무 좋을 것 같아요😊
게시판과 검색 기능은 떼어놓을 수 없겠죠...ㅎㅎ
검색기능은 제목, 도시, 사용자별로 검색할 수 있으면 좋을 것 같은데 시간이 부족할수도 있으니 제목 -> 도시 -> 사용자순으로 개발한다면 좋을 것 같습니다!

\+ 커뮤니티 기능은 아니지만 좋아요 누른 게시물을 모아서 볼 수 있다면 좋을 것 같아요! 시간이 남는다면...ㅎㅎ 생각해보죠...ㅎㅎ

**Comment by LJW25** [comment]
### 안건 1

저도 `3. 1번과 2번의 기능을 전부 제공한다.` 입니다.
다만 제공 방식은 디노의 의견과 조금 다른게, `비공개 → 링크공개 → 전체공개` 세단계로 이루어져야 한다고 생각해요.
왜냐면 `공개 - 비공개` `공유-비공유` 로 나눌 경우 `공개/비공유` 의 선택이 가능해지는데, 그럼 하나의 여행에 전체공개용 코드와 링크공개용 코드를 나눠서 운영해야하기 때문이죠.. 그래서 하나의 여행엔 하나의 코드로만 관리하고 해당 코드의 공유상태로 관리하는 편이 더 적절하다고 생각합니다. 
UX의 관점에서도 전체 공개가 된 게시글이 링크공유가 막힐 이유도 없다고 생각해요!

### 안건 2

상호작용
- 게시물 좋아요 기능
    - 로그인 한 사람만 가능
    - 내 게시글에 좋아요 한 사람을 볼 수 있다면? 너무 인스타같나?
- 게시물 조회수 표시 기능

조회
- 게시물 추천 기능 → 커뮤니티 메인 페이지
    - 최근 일주일간 조회수/좋아요가 많은 순
    - 새로 올라온 글 (중에 아이템이 10개 이상인 글) 등
- 좋아요 한 게시물 목록 조회 기능
- 게시물 검색 기능(제목별, 도시별, 작성자별)

---

이외에도 홍고가 언급한 신고기능도 좋을 것 같습니다. 커뮤니티가 생긴다면 이상한 사진을 올리는 게시글을 차단할 필요가 있으니까요! 
신고 들어오면 우리 행록계정으로 메일이 들어오게 한다던지 하면 좋을듯? 물론 필수 아니고 부가입니다.

**Comment by waterricecake** [comment]
### 안건 1
저도 `3. 1번과 2번의 기능을 전부 제공한다.`입니다.
디노의 의견처럼 `공개-비공개`, `공유-비공유`로 구현하는 것이 현재 코드를 유지하면서 추가할 수 있다고 생각합니다.
이상적으로는 인스타 처럼 팔로우 기능을 통해 친구공개로 하고 싶긴하지만....

### 안건 2
제가 생각한 부가 기능은
게시물 
- 좋아요 기능 (로그인 유저만)
- 신고 기능

있으면 좋겠다
- 댓글 기능
- 대댓글 기능

조회
- 검색기능 (사용자, 도시, 제목)
- 게시물 좋아요 순으로 정렬 ( 일별, 월별로 나누기)

있으면 좋겠당
- 어드민페이지 -> 신고등 관리

**Comment by mcodnjs** [comment]
### 안건1

> 1. 욕심을 내면 1,2,3 외의 제 4의 팔로우 기능!! => 인스타 구조와 동일

유저가 팔로우한 사람의 여행 기록을 볼 수 있음(~~팔로우 수락, 취소도 필요..~~) 
게시글의 공유 형태를 정하는 것이 아닌, **사용자 계정의 private/public 여부**가 될 듯!

제안 이유: 
- 여행 기록의 경우 private한 정보가 많음
- 전체 공개의 커뮤니티에 게시글을 업로드 하였을 때, 
수익성이 창출되지 않는 이상 이 서비스를 사용하는 유저들에게 전체 공개를 하는 것은 쉽지 않은 행위라고 생각
- 계정의 Private/Public을 넣을 수 있다면, 굳이 링크를 통해 서비스를 사용하지 않아도 되는 장점 -> 서비스 유입이 자연스럽게 많아짐
- 회원/미회원의 공유 니즈를 모두 충족시킬 수 있음


> 2. 현실적으로 본다면 링크 공개와 전체 공개 모두 제공하는 3번!

**게시글 공유의 종류**가 링크 공개, 커뮤니티 공개 두가지 형태로 제공
커뮤니티 공개를 한다면 우리의 커뮤니티에 해당 게시글이 업로드 되는 형태!

---

### 안건2
저는 만약 커뮤니티 기능이 생긴다면 있어야하는 기능을 우선순위로 적어봤습니다 ~! 
- 회원가입 시 작성자 닉네임 중복 체크 기능
- 검색 기능 (게시글, 작성자, 도시 등등) -> ~~SQL로 최대한 최적화를 해본 후 성능 저하가 느껴질 때 ES를 도입해야하지 않을까 ..~~
- 좋아요 > 조회수 > 댓글 (우선순위 순)
  - 좋아요가 있다면, 유저가 좋아요한 글을 모아보는 기능도 따라옴
  - 링크 공개한 글도 좋아요 할 수 있나?
- 인기글 (좋아요수와 조회수를 적절히 조합하여 산출?)
